runtime-sdk/src/modules/core/mod.rs

```diff
@@ -30,10 +42,103 @@ pub enum Error {
     #[error("insufficient balance to pay fees")]
     #[sdk_error(code = 5)]
     InsufficientFeeBalance,
+
+    #[error("invalid argument")]
+    #[sdk_error(code = 6)]
+    InvalidArgument,
+
+    #[error("gas overflow")]
+    #[sdk_error(code = 7)]
+    GasOverflow,
+
+    #[error("out of gas")]
+    #[sdk_error(code = 8)]
+    OutOfGas,
+}
+
+pub trait API {
+    fn use_gas(ctx: &mut TxContext<'_, '_>, gas: u64) -> Result<(), Error>;
 }
 
 /// State schema constants.
 pub mod state {
     /// Runtime metadata.
     pub const METADATA: &[u8] = &[0x01];
 }
+
+pub struct Module;
+
+const CONTEXT_KEY_GAS_USED: &str = "core.GasUsed";
+
+impl API for Module {
+    fn use_gas(ctx: &mut TxContext<'_, '_>, gas: u64) -> Result<(), Error> {
+        let gas_limit = ctx.tx_auth_info().fee.gas;
+        let gas_used = ctx.tx_value::<u64>(CONTEXT_KEY_GAS_USED);
+        let new_gas_used = match gas_used.overflowing_add(gas) {
+            (new_gas_used, false) => new_gas_used,
+            (_, true) => return Err(Error::GasOverflow),
+        };
+        if new_gas_used > gas_limit {
+            return Err(Error::OutOfGas);
+        }
+        *gas_used = new_gas_used;
+        Ok(())
+    }
+}
+
+impl Module {
+    fn query_estimate_gas(
+        ctx: &mut DispatchContext<'_>,
+        args: transaction::Transaction,
+    ) -> Result<u64, Error> {
+        let mi = ctx
+            .methods
+            .lookup_callable(&args.call.method)
+            .ok_or(Error::InvalidMethod)?;
+        let mut sim_ctx = DispatchContext {
+            mode: Mode::SimulateTx,
+            runtime_header: ctx.runtime_header,
+            runtime_round_results: ctx.runtime_round_results,
+            runtime_storage: ctx.runtime_storage,
+            io_ctx: ctx.io_ctx.clone(),
+            methods: ctx.methods,
+            max_messages: ctx.max_messages,
+            messages: Vec::new(),
+            values: BTreeMap::new(),
+        };
```

Comment:
This should probably be factored out into a `with_simulation` method on the context.

Reply:
ok